File: Java/Beginner/D_GraphSearch/boj_24445.java
package Beginner.D_GraphSearch;

import java.io.*;
import java.util.*;

public class boj_24445 {
    static List<Integer>[] grp;
    static StringTokenizer st;
    static final StringBuilder sb = new StringBuilder();
    static final BufferedReader br =
            new BufferedReader(new InputStreamReader(System.in));

    private static void solve(int r) {
        Queue<Integer> q = new LinkedList<>();
        boolean[] visited = new boolean[grp.length];

        q.add(r);
        visited[r] = true;

        int[] ans = new int[grp.length];
        int cnt = 1;

        while (!q.isEmpty()) {
            int now = q.poll();
            ans[now] = cnt++;

            for (int nxt : grp[now]) {
                if (!visited[nxt]) {
                    visited[nxt] = true;
                    q.add(nxt);
                }
            }
        }

        for (int i = 1; i < ans.length; i++)
            sb.append(ans[i]).append("\n");

        System.out.print(sb);
    }

    public static void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int r = Integer.parseInt(st.nextToken());

        grp = new ArrayList[n + 1];
        Arrays.setAll(grp, elem -> new ArrayList<>());

        while (m-- > 0) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());

            grp[u].add(v);
            grp[v].add(u);
        }

        for (var g : grp)
            Collections.sort(g, Collections.reverseOrder());

        solve(r);

        br.close();
    }
}

File: Kotlin/src/main/kotlin/beginner/d_GraphSearch/boj_24445.kt
package beginner.d_GraphSearch

import java.util.*

private fun solve(grp: Array<MutableList<Int>>, start: Int) {
    val deque = ArrayDeque<Int>()
    val visited = Array<Boolean>(grp.size) { false }
    val ans = Array<Int>(grp.size){ 0 }
    var cnt = 1

    deque.add(start)
    visited[start] = true

    while (deque.isNotEmpty()) {
        val now = deque.pollFirst()
        ans[now] = cnt++

        for (next in grp[now]) {
            if (!visited[next]) {
                visited[next] = true
                deque.add(next)
            }
        }
    }

    for (i in 1 until ans.size) println(ans[i])
}


fun boj_24445() = with(System.`in`.bufferedReader()) {
    val (n, m, r) = readLine()
        .split(" ")
        .map { it.toInt() }

    var grp: Array<MutableList<Int>> = Array(n + 1) { mutableListOf() }

    repeat(m) {
        val (u, v) = readLine()
            .split(" ")
            .map { it.toInt() }

        grp[u].add(v)
        grp[v].add(u)
    }

    for (g in grp) g.sortDescending()
    solve(grp, r)
}